tools/ctexplain/types.py

```diff
@@ -0,0 +1,93 @@
+# Lint as: python3
+# Copyright 2020 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""The core data types ctexplain manipulates."""
+
+from typing import Dict
+from typing import Optional
+from typing import Tuple
+from dataclasses import dataclass
+from dataclasses import field
+
+
+@dataclass(frozen=True)
+class Configuration():
+  """Stores a build configuration as a collection of fragments and options."""
+  # BuildConfiguration.Fragments in this configuration, as base names without
+  # packages. For example: ["PlatformConfiguration", ...].
+  fragments: Tuple[str, ...]
+  # Dict of FragmentOptions to option key/value pairs. For example:
+  # {"CoreOptions": {"action_env": "[]", "cpu": "x86", ...}, ...}.
+  #
+  # Option values are stored as strings of whatever "bazel config" outputs.
+  #
+  # Note that Fragment and FragmentOptions aren't the same thing.
+  options: Dict[str, Dict[str, str]]
+
+  def __hash__(self):
+    return self._hash_value()
+
+  def __eq__(self, other):
+    return other._hash_value() == self._hash_value()
```

Comment:
Maybe a comment that this is susceptible to collisions?

I know we do this internally; however, there the compared hashes are longer.

Reply:
Keeping this pending on our discussion about exploring alternatives to `__hash__` / `__eq__`.

Reply:
Re-wrote data classes using frozendict from https://pypi.org/project/frozendict/, as we discussed.